docs/enhancements/feat-add-ability-to-tune-azuredisk-performance-parameters.md

```diff
@@ -0,0 +1,93 @@
+  - [Scope Of the Change](#scope-of-the-change)
+- [Limitations](#limitations)
+- [Future Considerations](#future-considerations)
+<!-- /toc -->
+
+## Summary
+
+Persistent volumes in kubernetes are used for wide variety of stateful workloads.
+These workloads have different runtime/IO characterstics, certain device level config settings
+on the data disk can make a huge difference in performance of the application.
+
+Azure storage publishes [guidelines](https://docs.microsoft.com/en-us/azure/virtual-machines/premium-storage-performance)
+for the applications to configure the disks' guest OS settings to drive maximum IOPS and Bandwidth.
+
+Azure Disk CSI driver users currently do not have an easy way to tune disk device configuration to
+get better performance out of them for their workloads.
+
+This proposal proposes to provide users with ability to enable automatic perf optimization of data
+disks by tweaking guest OS level device/disk/driver parameters.
+
+## Motivation
+
+As the adoption of the Azure Disk CSI driver increases, we will encounter different type
+of workloads which have different disk IO chracterstics. Some may require to drive the data
+disks to maximum IOPS, while others may need to do larger size writes and drive maximum throughput.
+
+Azure Disk CSI driver should evolve to let users tune the data disks configurations, to get optimal
+performance for their workloads.
+
+With this feature, we try to provide a configurable and handsfree way for the users to enable
+perf optimizations of data disk by enabling a feature in storageclass and be able to select from
+multiple optimization profiles based on their requirement.
+
+We intend to provide a way for the users to opt-in for the new behavior and expect not to break
+any existing applications/configurations.
+
+## Proposal
+
+Azure Disk CSI driver will read two extra flags from the storageclass, perftuningmode and perfProfile.
+
+- *perftuningmode*: Can be set to `Auto` or `None`. In `Auto` mode, CSI driver will automatically
+calculate the optimal configuration for the disk. Users will be able to control some characterstics of the
+optimization by selecting different `perfprofile`. The driver defaults to `None`, which means no optimization
+is done. Please read the limitations to fully understand the options available today.
+- *perfprofile*: Users will be able to select a certain performance profile for their device to match their
+workload needs. For ex: we will expose different profile options to let users configure the disks for balanced
+IO, high throughput or high IOPS etc. Please read the limitations to fully understand the options available today.
+
+### Tuning Mode And Profile
+
+```yaml
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: sc-kubestone-perf-optimized-premium-ssd-csi
+provisioner: disk.csi.azure.com
+parameters:
+  skuname: Premium_LRS
+  perftuningmode: Auto
+  perfprofile: Default
```

Comment:
why not combine these two parameters, e.g. only have `perfTuningMode`, now it has two available values: ``(empty means none, by default), `auto`, it could save one parameter!

Reply:
@edreed @andyzhangx has an interesting idea. I know we have discussed this in the past. We can just have perf tuning mode, which can be a manual or auto mode. We will validate this in code. I think from what I was planning I can't think of a scenario where it would not work. Let me know what you think?

Reply:
I agree that these can be combined into a single `perfTuningMode`. I would recommend supporting the values `none` and `auto`. The default would be `none` and implied if the parameter was not present. We could add additional properties in the future to support manual or profile based tuning.

Reply:
@edreed I was more inclining towards providing the name of the profiles directly and that profile can be an inbuilt one or something provided by customer using a CRD. absence would mean no profile provided, which is same as tuning not enabled. so the current implementation will change to:
* perfprofile not present: tuning disabled
* perfprofile present but is empty or not known (today only default will be known): will result in failure.
* perfprofile present and known (today only default will be known): we will apply the profile. known profile will be combination of inbuilt profiles and customer provided profiles.

What do you think?

Reply:
@abhisheksinghbaghel Yes, that would work and be extensible without additional properties in the future. I believe that we should also allow for a `none` profile to be specified in addition to omitting the parameter to allow the config to be explicit in perf tuning choice.

Reply:
@edreed Thanks. I can make that happen!

Reply:
so what's the conclusion? use two parameters? and what's the default value for this driver? enable or disable?

Reply:
One param. feature will be disabled by default, with a value of perfTuningMode=None